api/graphite.go

```diff
@@ -891,19 +891,22 @@ func (s *Server) graphiteTagFindSeries(ctx *middleware.Context, request models.G
 
 func (s *Server) clusterFindByTag(ctx context.Context, orgId uint32, expressions []string, from int64, maxSeries int) ([]Series, error) {
 	data := models.IndexFindByTag{OrgId: orgId, Expr: expressions, From: from}
-	responseChan, errorChan := s.peerQuerySpeculativeChan(ctx, data, "clusterFindByTag", "/index/find_by_tag")
+	newCtx, cancel := context.WithCancel(ctx)
+	responseChan, errorChan := s.peerQuerySpeculativeChan(newCtx, data, "clusterFindByTag", "/index/find_by_tag")
 
 	var allSeries []Series
 
 	for r := range responseChan {
 		resp := models.IndexFindByTagResp{}
 		_, err := resp.UnmarshalMsg(r.buf)
 		if err != nil {
+			cancel()
```

Comment:
Would it make sense to just `defer cancel()` at the top like the peerQuery functions do? 

Also, how much does this canceling really buy us if the top-level ctx will be `Done` when the error is returned?

Reply:
not sure what you mean with a ctx "being Done".
`ctx.Done()` just returns a channel that signals cancellation, but you still need to call a corresponding cancel function for that to happen.

I see 3 callers of this function:
* `/render` -> `renderMetrics` -> `executePlan`
* `/tags/findSeries` (graphiteTagFindSeries)
* `querier.Select`

we would need all 3 to call a cancelfunc for my change to be unnecessary.
but from what i can tell, it looks like neither of them, or at least not the first two, do this.

as for using defer vs just putting the code at the return sites, for short functions it's really debatable and depends more on taste. so unless you have a strong argument, let's just stick with it.

Reply:
the cancels that i added in clusterFindByTag target cases that `peerQuerySpeculativeChan` cannot detect itself. these are : 
* unmarshalling error
* exceeding maxSeriesPerReq

the cancels triggered in `peerQuerySpeculativeChan`  (upon erroring peer.Post and upon function return) would be triggered not all, or much later, respectively.

Reply:
> the parent context's Done channel will be closed 

where does this happen?

Reply:
TBH, I didn't come to this conclusion by code inspection, but rather by logging when early responses happened. It definitely was how it behaved, but I couldn't tell you if it's by contract or a symptom of how I was querying

Reply:
I see, well. this reinforces my point that these matters should be made much more obvious. hence rather a call to cancel more rather than too few.
good to merge?

Reply:
Yeah, let me rebase